### PR TITLE
DM-34865: Script for looking at ap_verify metrics

### PR DESCRIPTION
**Pull request by parejkoj**

### CODE REVIEW

**Review by kfindeisen — APPROVED** [review]
Looks good! Mostly style and documentation comments, and a few feature requests for difference mode.

File: doc/lsst.verify/index.rst [hunk 1]
```diff
@@ -84,6 +84,7 @@ Script reference
    scripts/gen3_to_job.py
    scripts/inspect_job.py
    scripts/lint_metrics.py
+   scripts/print_metricvalues
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply doc/lsst.verify/index.rst:87]: No; following [RFC-698](https://jira.lsstcorp.org/browse/RFC-698), I made the script not have an extension. The docs build properly.

[the remark this answers is not in the document]
Comment (kfindeisen) [reply doc/lsst.verify/index.rst:87]: But the script is called `print_metricvalues.py`. Have you tried running `print_metricvalues ...` as recommended by the docs? It won't work.

File: doc/lsst.verify/scripts/print_metricvalues.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.. autoprogram:: lsst.verify.bin.print_metricvalues:build_argparser()
+   :prog: print_metricvalues
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply doc/lsst.verify/scripts/print_metricvalues.rst:2]: I don't understand your comment here? The built docs that I see look correct.

File: python/lsst/verify/bin/print_metricvalues.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+from .. import extract_metricvalues
+
+
+def build_argparser():
+    """Return an ArgumentParser for this script.
+    """
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        epilog='More information is available at https://pipelines.lsst.io.')
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/bin/print_metricvalues.py:39]: That comes from the example template: https://developer.lsst.io/stack/argparse-script-topic-type.html#the-autoprogram-directive

File: doc/lsst.verify/index.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply doc/lsst.verify/index.rst:87]: Woah! I must have failed to push the rename early on, and I always rely on tab-completion so I didn't even notice. Thanks for catching that. I'm vaguely surprised the docs built without error though!

File: python/lsst/verify/bin/print_metricvalues.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                                           data_id_keys=args.data_id_keys,
+                                           verbose=args.verbose)
+    else:
+        collection2 = args.collection2 if args.collection2 is not None else args.collection
+        butler2 = lsst.daf.butler.Butler(args.repo2, collections=collection2)
+        print(f"Difference between {args.repo}#{args.collection} and {args.repo2}#{collection2}")
+        extract_metricvalues.print_diff_metrics(butler,
+                                                butler2,
+                                                data_id_keys=args.data_id_keys,
+                                                verbose=args.verbose)
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/bin/print_metricvalues.py:85]: Doing deltas on time and memory seemed not very useful, given how variable they can be; the main point of the deltas was to see if a metric changed with a code change, without uploading to chronograph. Adding them could be a future feature, if people want it.

File: python/lsst/verify/extract_metricvalues.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tools for loading metric values from a butler and printing them, or from
+two butlers and differencing them.
+"""
+__all__ = ["load_value", "load_timing", "load_memory",
+           "print_metrics", "print_diff_metrics", "load_from_butler"]
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py:26]: Yes, that was intentional. They seem like something that would be quite useful to use from e.g. a notebook, to load a block of metrics in for further processing.

File: python/lsst/verify/extract_metricvalues.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        case "timing":
+            result = load_timing(butler, verbose=verbose)
+
+            def value_formatter(value):
+                return f"{value.datum.label}: {value.quantity:.4}"
+        case "memory":
+            result = load_memory(butler, verbose=verbose)
+
+            def value_formatter(value):
+                return f"{value.datum.label}: {value.quantity.to(u.Mbyte):.5}"
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py]: Ooof, you're right. I always forget that.

File: python/lsst/verify/extract_metricvalues.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    failed = 0
+    for key in result1:
+        data_id, metric = key
+        try:
+            value1 = result1[key]
+            value2 = result2[key]
+        except KeyError:
+            print(f"Result 2 does not contain metric '{metric}'' for dataId '{data_id}'")
+            failed += 1
+            continue
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py]: Yeah, I decided to punt on some of the error handling, because I only have a single use case for this right now. I don't know that it's worth expanding that until we have some more uses?

File: python/lsst/verify/extract_metricvalues.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        keys2 = sorted(list(result2.keys()))
+        print()
+        print(f"butler1 metrics found: {len(result1)}")
+        print(f"butler2 metrics found: {len(result2)}")
+        print(f"metrics in butler1 that were not found in butler2: {failed}")
+        print("Check that the butler schemas are comparable, if you expected one-to-one comparability.")
+        if verbose:
+            print("Full DataCoordinates for the first key of each result, to compare schemas:")
+            print(keys1[0][0].full)
+            print(keys2[0][0].full)
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py:153]: butler registry schemas determine whether you can actually compare metrics in different repos. I discovered this when trying to compare different `ap_verify` runs, one before the `visit_system` registry change, and one after. It reported that nothing was found, because the dataIds weren't comparable. Instead of trying to hack together a way to make them comparable (it's not at all easy, per slack discussion), I just added this message.

See more here: https://lsstc.slack.com/archives/C01FBUGM2CV/p1654727269705119

File: python/lsst/verify/extract_metricvalues.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+            value1 = result1[key]
+            value2 = result2[key]
+        except KeyError:
+            print(f"Result 2 does not contain metric '{metric}'' for dataId '{data_id}'")
+            failed += 1
+            continue
+
+        delta = value1.quantity - value2.quantity
+        if delta != 0 or verbose:
+            print(f"{_data_id_label(data_id, data_id_keys)};", value1.datum.label, delta)
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py]: I think whether something is significant depends very much on the use case: sometimes the delta should be identically zero everywhere. I've made it print `{delta} / {value1.quantity}`, which at least gives the reference.

File: python/lsst/verify/extract_metricvalues.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        Butler to load values from.
+    verbose : `bool`, optional
+        Print extra information when loading values.
+
+    Returns
+    -------
+    result : `dict` [`tuple`, `MetricValue`]
+        The loaded metric values, keyed on data_id and metric.
+    """
+    return load_from_butler(butler, "metricvalue*", reject_suffix=("Time", "Memory"), verbose=verbose)
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py:181]: I'm just going off the ap_verify metrics: I don't know if we actually have a convention for other things. If we don't we'll have to figure out another way to separate them.

File: python/lsst/verify/bin/print_metricvalues.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/verify/bin/print_metricvalues.py:39]: If it's just there as example fluff, I'd say leave it out...

File: python/lsst/verify/bin/print_metricvalues.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/verify/bin/print_metricvalues.py:85]: All I'm asking is that you document it so that users aren't surprised.

File: python/lsst/verify/extract_metricvalues.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/verify/extract_metricvalues.py:153]: I'm asking about the opposite, though -- if a metric value dataset is found in one repository but not another, why would you conclude that it's a schema change instead of the dataset really not being there?

File: python/lsst/verify/bin/print_metricvalues.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/bin/print_metricvalues.py:39]: I've asked about this on #dm-docs-support on slack.

File: python/lsst/verify/extract_metricvalues.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py:153]: I'm not concluding that, it's just something the user has to check: it really confused me when nothing was found. I've reworded the message a bit.

Not having any more usecases than ap_verify, I don't really know what the corner cases here are supposed to be. I'd rather not worry about those things until this gets some broader use.

File: tests/data/ci_hits2015-DM-34786/ap_verify-output/20220603T234025Z/metricvalue_ap_association_AssociationTime/20150218/g/g_DECam_SDSS_c0001_4720.0_1520.0/411371/metricvalue_ap_association_AssociationTime_DECam_g_g_DECam_SDSS_c0001_4720_0_1520_0_411371_N25_ap_verify-output_20220603T234025Z.yaml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+      unit: ''
+      value: '2022-06-03T23:44:58.535135'
+  identifier: ac94ca2fa4b54a43b614b3334919454d
+  name: ap_association.AssociationTime
+identifier: a5870a991cec4c3bba8e278df3a6a357
+metric: ap_association.AssociationTime
+notes:
+  ap_association.AssociationTime.estimator: utils.timer.timeMethod
+unit: s
+value: 0.0065764230000002755
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply tests/data/ci_hits2015-DM-34786/ap_verify-output/20220603T234025Z/metricvalue_ap_association_AssociationTime/20150218/g/g_DECam_SDSS_c0001_4720.0_1520.0/411371/metricvalue_ap_association_AssociationTime_DECam_g_g_DECam_SDSS_c0001_4720_0_1520_0_411371_N25_ap_verify-output_20220603T234025Z.yaml:22]: We made a set of fake metrics, which helped identify some bugs in the metricvalue loading code.

File: python/lsst/verify/bin/print_metricvalues.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/bin/print_metricvalues.py:39]: The word from Jonathan is: "My intent with that advice was to be a helpful way of connecting a random executable that someone might find with the larger Science Pipelines project so that could understand where it came from."

File: python/lsst/verify/bin/print_metricvalues.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                        help="Only print these dataId keys in the output;"
+                             "for example, `--data_id_keys detector visit`.")
+
+    parser.add_argument("--data_id_restriction", "-r",
+                        type=lambda kv: kv.split("=", 1), action="append",
+                        default=None,
+                        help="Restrict output to dataIds matching these `key=value` pairs; "
+                             "for example, `-r detector=60`."
+                             "TODO: This doesn't behave how I want: should I bother to implement it"
+                             "on the cmdline or just drop it?")
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/bin/print_metricvalues.py:62]: I've removed this cmdline arg.

File: python/lsst/verify/extract_metricvalues.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+        Print extra information when loading.
+
+    Returns
+    -------
+    result : `dict` [`tuple`, `MetricValue`]
+        The loaded metric values, keyed on data_id and metric.
+    """
+    # all possible metrics that have been registered
+    metrics = butler.registry.queryDatasetTypes(query)
+    data_ids = butler.registry.queryDataIds(("detector", "visit"), datasets="calexp")
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py]: The test metrics we made helped sort this out, and this code has been substantially refactored.

File: tests/test_extract_metricvalues.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            """
+            with contextlib.redirect_stdout(io.StringIO()) as stdout:
+                extract_metricvalues.print_metrics(butler, kind, **kwargs)
+                # Split up the lines, and remove empty ones.
+                result = list(filter(None, stdout.getvalue().split("\n")))
+                self.assertEqual(len(result), n)
+                self.assertEqual(last_line, result[-1])
+
+        # default call with no kwargs
+        expect = "{instrument: 'DECam', detector: 60, visit: 411371, ...}; pipe_tasks.numDeblendedSciSources: 125.0"  # noqa: E501
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply tests/test_extract_metricvalues.py]: I don't know about dataId stability. I think we'll deal with that if it ever becomes an issue.

File: python/lsst/verify/extract_metricvalues.py [hunk 15]
```diff
@@ -107,23 +114,29 @@ def print_diff_metrics(butler1, butler2, data_id_keys=None, verbose=None):
...
+        if old_data_id != data_id:
+            print(f"\n{_data_id_label(data_id, data_id_keys)}")
+            old_data_id = data_id
+
         try:
             value1 = result1[key]
             value2 = result2[key]
         except KeyError:
-            print(f"Result 2 does not contain metric '{metric}'' for dataId '{data_id}'")
+            print(f"Result 2 does not contain metric '{metric}'")
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py:128]: The dataId is now printed at the top, so it's redundant.

File: tests/test_extract_metricvalues.py [hunk 16]
```diff
@@ -172,12 +181,16 @@ def check_stdout(n, last_line, **kwargs):
             return result
 
         last_line = "Number of metrics that are the same in both runs, out of 53: 28"
-        result = check_stdout(26, last_line)
-        expect = "{instrument: 'DECam', detector: 60, visit: 411371, ...}; ap_association.numNewDiaObjects -513.0 ct"  # noqa: E501
+        result = check_stdout(32, last_line)
+        expect = "{instrument: 'DECam', detector: 60, visit: 411371, ...}"
+        self.assertIn(expect, result)
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply tests/test_extract_metricvalues.py:186]: I could implement something for that, but for the difference output I want to test multiple strings and it's not worth the bother.

File: python/lsst/verify/extract_metricvalues.py [hunk 17]
```diff
@@ -107,23 +114,29 @@ def print_diff_metrics(butler1, butler2, data_id_keys=None, verbose=None):
 
     same = 0
     failed = 0
+    old_data_id = None
     for key in result1:
         data_id, metric = key
+        if old_data_id != data_id:
+            print(f"\n{_data_id_label(data_id, data_id_keys)}")
+            old_data_id = data_id
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/verify/extract_metricvalues.py]: I tried to do this, but gave up; we'll live with the duplication for now.

**Comment by parejkoj** [comment]
@kfindeisen : take another look if you want to: I refactored a lot of it. Thanks for your help along the way, I think it's much more robust now.
